### PR TITLE
Add alt text for gcc-tips

### LINKED ISSUE #4028: [docs] Add alt text on images
## Summary

For every image in LightGBM's documentation, alternative text that describes the image should be added.

## Motivation

People with limited vision use screen readers to view web pages. When these screen readers encounter an image, they will only read some placeholder like "IMAGE" by default. However, if [the "alt" tag](https://www.w3schools.com/tags/att_img_alt.asp) is specified, many screen readers will use that instead.

Adding alt text for images would improve the accessibility of LightGBM's documentation for users who use screen readers.

## Description

To help with this issue, submit a pull request that adds alt text for images in LightGBM's documentation. Please limit pull requests to only 1 file, to simplify reviews.

As of this writing, the following documentation files have images (based on `git grep -E "\.\. image" docs`)

* [x] `docs/Features.rst` (#4036)
* [x] `docs/GPU-Performance.rst` (#4036)
* [x] `docs/GPU-Windows.rst` (#4036)
* [x] `docs/Parallel-Learning-Guide.rst` (#4030)
* [x] `docs/Parameters-Tuning.rst` (#4035)
* [x] `docs/gcc-Tips.rst` (#4187)
* [x] `docs/index.rst` (#4029)

When adding alt text, follow the advice in https://accessibility.psu.edu/images/alttext/.

> ...include what you might relay over the phone

> it should describe the image. It should NOT include information about copyright or any extra information that would be relevant to a seeing audience.

This can be done with the `:alt:` parameter in Sphinx `..image` directives (see [the Sphinx docs](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/images.html) for details).

For example:

```rst
.. image:: ./_static/images/some-image.png
  :alt: A screenshot with build logs from GCC.
```

## References

While working on this, you might also consider some of the recommendations from https://webaim.org/techniques/screenreader/#how. For example, things that should be read out loud as acronyms, like `gcc`, should be spelled out all in capital letters ("GCC").

### PR DESCRIPTION
**Pull request by akshitadixit**
Closes #4028 

Added alt text for images in gcc-tips.rst, sorry for taking this long, was reading about things since I did not fully understand all of the images. Kindly review and suggest in case I missed on something.

### CODE REVIEW

File: docs/gcc-Tips.rst [hunk 1]
```diff
@@ -32,27 +32,34 @@ Some explanatory pictures:
...
 
 .. image:: ./_static/images/gcc-chart.png
    :align: center
    :target: ./_static/images/gcc-chart.png
+   :alt: this picture is a rendition of first where y-axis is cost and x-axis is number of threads comparing the four algorithms
 
 .. image:: ./_static/images/gcc-comparison-1.png
    :align: center
    :target: ./_static/images/gcc-comparison-1.png
+   :alt: a bar chart comapring Light G.B.M performance versus compilation flags which is maximum at depth 8 with Light G.B.M version 2, flag O 3
```
Comment (jameslamb) [inline docs/gcc-Tips.rst:50]:
```suggestion
   :alt: a horizontal bar chart comparing Light G B M performance versus compilation flags. For most settings of max depth, best performance was achieved with flags dash O 3, dash M T U N E equals native.
```

I think the description for this one needs to be reworked. I don't think it's important to mention that the best performance on this particular problem was achieved with `max_depth=8`, since that's highly dependent on the dataset. The most important conclusion from this plot is that you should expect compiling with `-O3 -mtune=native` to give the best performance results for different settings of `max_depth`.

File: docs/gcc-Tips.rst [hunk 2]
```diff
@@ -32,27 +32,34 @@ Some explanatory pictures:
...
 
 .. image:: ./_static/images/gcc-comparison-2.png
    :align: center
    :target: ./_static/images/gcc-comparison-2.png
+   :alt: a bar chart comapring Light G.B.M cumulative performance versus compilation flags which is maximum at depth 8 with Light G.B.M version 2, flag O 3
```
Comment (jameslamb) [inline docs/gcc-Tips.rst:55]:
```suggestion
   :alt: a set of 4 vertical bar charts comparing Light G B M performance versus compilation flags. For most settings of max depth, best performance was achieved with flags dash O 3, dash M T U N E equals native.
```

I think this could have the same description as the image above it, just with "horizontal" switched to "vertical".

File: docs/gcc-Tips.rst [hunk 3]
```diff
@@ -32,27 +32,34 @@ Some explanatory pictures:
...
 
 .. image:: ./_static/images/gcc-meetup-1.png
    :align: center
    :target: ./_static/images/gcc-meetup-1.png
+   :alt: comparison of cumulative speed versus the slowest algorithm at various depths with v-2, O-3 remaining constant almost in all cases
```
Comment (jameslamb) [inline docs/gcc-Tips.rst:60]:
```suggestion
   :alt: grid of line charts showing the relative speed of training for different combinations of max depth, number of threads, and compilation flags. The grid shows that for models with max depth greater than 5, compiling Light G B M with default compiler flags produces faster training time than any of the customizations explored.
```

I think the story here is a little different. Will you consider this re-write?

File: docs/gcc-Tips.rst [hunk 4]
```diff
@@ -32,27 +32,34 @@ Some explanatory pictures:
...
 
 .. image:: ./_static/images/gcc-meetup-2.png
    :align: center
    :target: ./_static/images/gcc-meetup-2.png
+   :alt: comparison of cumulative speed versus the slowest density of each algorithm at various depths with v-2, O-3 remaining constant almost in all cases
```
Comment (jameslamb) [inline docs/gcc-Tips.rst:65]: I think these stacked area plots (https://lightgbm.readthedocs.io/en/latest/_static/images/gcc-meetup-2.png) are really difficult to understand, and that this one should just be removed.

@StrikerRUS what do you think?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thank you so much for taking this on @akshitadixit ! This page has received very little attention in the 3+ years since this content was first introduced, and I understand that it's pretty difficult to interpret some of the plots there, which don't have any explanatory text. This is the first time I've actually read this page and even I was pretty confused by some of it.

I left some suggestions below, but we'll also need to wait for another maintainer to comment on my questions about possibly removing some of these plots.

By the way...I noticed that you made this PR from the `master` branch on your fork. Since you're planning to help us with other things (https://github.com/microsoft/LightGBM/issues/4136#issuecomment-819274063), in the future you'll probably find it useful to create a new branch on your fork for each pull request. That way, you can have multiple pull requests into this project active at the same time.

**Comment by StrikerRUS** [comment]
`gcc Tips` is @Laurae2's baby: #511, https://github.com/microsoft/LightGBM/pull/945#issuecomment-334923417. Maybe he would like to comment or update this doc.

File: python-package/lightgbm/sklearn.py [hunk 5]
```diff
@@ -76,7 +76,7 @@ def __call__(self, preds, dataset):
         elif argc == 3:
             grad, hess = self.func(labels, preds, dataset.get_group())
         else:
-            raise TypeError("Self-defined objective function should have 2 or 3 arguments, got %d" % argc)
+            raise TypeError(f"Self-defined objective function should have 2 or 3 arguments, got {argc}")
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:79]: Can you please remove these unrelated changes from this PR? I think they were meant for https://github.com/microsoft/LightGBM/pull/4181.

On whichever branch you use (based on my recommendation in https://github.com/microsoft/LightGBM/pull/4181), you can remove them with a command like `git reset --soft HEAD~1` followed by a force push.

Let me know if you get stuck!

**Comment by akshitadixit** [comment]
What a relief to know I am not the only person who was confused after seeing the page. I almost gave up on the PR afraid to pull naive alt texts to such a project. I read your suggestions and realized what I misinterpreted. Thank you!

**Comment by jameslamb** [comment]
> What a relief to know I am not the only person who was confused after seeing the page. I almost gave up on the PR afraid to pull naive alt texts to such a project. I read your suggestions and realized what I misinterpreted. Thank you!

Yes, sorry about that! I think that page could use some work. Thanks for pushing through it! Without your PR we probably would have not noticed that it needed some next explanatory text.

**Comment by jameslamb** [comment]
ok @akshitadixit I have an update for you! I talked to @Laurae2 in another chat about this. Copying those comments below:

> * first pic = with different depth hyperparameter values and different number of threads, the difference between using -O2 -mtune=core2, -O2 -mtune=native, -O3 -mtune=native, and -O3 -fast-math -mtune=native is negligible, although -mtune=native is slightly faster (+2%)
> * second pic = same as first pic (smaller bar = faster, it is cumulated time)
> * third pic = -O2 -mtune=core2 mean time is larger than the others (hence slower), -O3 -mtune=native is the fastest
> * fourth pic = using -O2 -mtune=core2 as the baseline, all 3 other options are faster, with -O3 -mtune=native being the fastest
> * fifth pic = same as fourth pic, but chart is grouped by compiler set of options instead of depth
> * sixth pic = overall all performed tests, how much cumulated time (in percentage) you would save by using which compiler set of options
> * seventh pic = against overall possible time saved, how much each compiler set of options will save you in percentage

> the pics can be deleted since the most important is to tell gcc users to use -O3 -mtune=native and link to the experiment (the pics are only the details to not have to sift through 100+ slides, if really need a pic then 5th pic is best to keep


So given that, could you please change this pull request to remove all of the images EXCEPT for `gcc-comparison-2.png`? 

Sorry that you spent time on them and we're going to delete them but know that we wouldn't have made this change unless you brought it up! So this improvement of the docs is thanks to your efforts.

**Comment by akshitadixit** [comment]
Closing this to fix all of the mess I created :sweat_smile:
